### ISSUE TITLE
[Bug] Constant 403 errors while trying to load videoplayback

### ISSUE THREAD

**Issue by TheEdgeOfRage** [iv-org/invidious#4045]
**Describe the bug**
I host my own invidious instance and some videos just won't load on it. The thumbnail loads, but the video itself does not. There's just a spinner and the timeline says 0:00/0:00. This does not happen with all videos, but I haven't spotted a pattern yet.

When looking at the network tab, I can see that the browser is constantly sending requests to `/videoplayback` and getting a 403 status with no response body. It sends about 10 requests per second and doesn't stop until I close the tab. For reference, I use DASH on my instance.

**Steps to Reproduce**

1. Go to https://vid.theedgeofrage.com/watch?v=RBo1Mq1JMh8
2. Try to play video

**Logs**
Logs don't show anything besides the failing request log itself:

```
invidious-invidious-1     | 2023-08-10 09:50:50 UTC [info] 403 GET /videoplayback?expire=1691682024&ei=iLDUZJnwFNCOx_APssW3qAM&ip=94.130.141.40&id=o-ALEojr5Ke8EW5dc7Xuq-3wksylUWHvc81JCGu0Jw5dkp&itag=136&source=youtube&requiressl=yes&mh=Co&mm=31%2C26&mn=sn-4g5e6nzz%2Csn-f5f7kn7z&ms=au%2Conr&mv=m&mvi=4&pl=22&initcwndbps=445000&spc=UWF9f3uYf0Q_K1VEYIgDmPBkGM6dxrZqXRGhYOsRhw&vprv=1&svpuc=1&mime=video%2Fmp4&gir=yes&clen=438730931&dur=3145.775&lmt=1691291183341453&mt=1691660003&fvip=2&keepalive=yes&fexp=24007246&beids=24350018&c=ANDROID&txp=3309224&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIgLzQ6H_s28Scvwg6naQLHHEGK_-68OgYfih7J5xXWEdwCIQDdIlh8g6r6ZjTwl-kV2p4qAlFpOS0K6wzm0WBKq4wJsA%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIhAOBrXV03okr0vNFfd6sxZ-7xvwhPg1UNe209RUtr9O_wAiBaZWVNy_C4QhxWpOxZ1IP67fbfVGc03skcndJdI_ef2Q%3D%3D&host=rr4---sn-4g5e6nzz.googlevideo.com 50.29ms
```

**Screenshots**

![invidious](https://github.com/iv-org/invidious/assets/29902467/e59fa1c6-ac4c-4948-bce2-474f2d3a8999)

**Additional context**
 - Browser: Firefox 116
 - Desktop OS: Linux
 - Invidious deployed through docker, latest version from 2023-08-07
 
 docker-compose.yml file:
 
 ```yaml
 version: "3"
services:
  invidious:
    image: quay.io/invidious/invidious:latest
    restart: unless-stopped
    ports:
      - "127.0.0.1:8030:3000"
    environment:
      # Please read the following file for a comprehensive list of all available
      # configuration options and their associated syntax:
      # https://github.com/iv-org/invidious/blob/master/config/config.example.yml
      INVIDIOUS_CONFIG: |
        db:
          dbname: invidious
          user: kemal
          password: REDACTED
          host: invidious-db
          port: 5432
        check_tables: true
        hmac_key: REDACTED
        external_port: 443
        domain: vid.theedgeofrage.com
        https_only: true
        statistics_enabled: false
        use_quic: true
        popular_enabled: false
        captcha_enabled: false
        admins: [TheEdgeOfRage]
        channel_refresh_interval: 10m
        dark_mode: dark
        feed_menu: ["Subscriptions", "Playlists"]
        default_home: Subscriptions
        comments: ["youtube", "reddit"]
        quality: dash
        quality_dash: 1440p
        unseen_only: true
    healthcheck:
      test: wget -nv --tries=1 --spider http://127.0.0.1:3000/api/v1/comments/jNQXAC9IVRw || exit 1
      interval: 30s
      timeout: 5s
      retries: 2
    logging:
      options:
        max-size: "1G"
        max-file: "4"
    depends_on:
      - invidious-db

  invidious-db:
    image: docker.io/library/postgres:14
    restart: unless-stopped
    volumes:
      - postgresdata:/var/lib/postgresql/data
      - ./config/sql:/config/sql
      - ./docker/init-invidious-db.sh:/docker-entrypoint-initdb.d/init-invidious-db.sh
    environment:
      POSTGRES_DB: invidious
      POSTGRES_USER: kemal
      POSTGRES_PASSWORD: REDACTED
    healthcheck:
      test: ["CMD-SHELL", "pg_isready -U $$POSTGRES_USER -d $$POSTGRES_DB"]

volumes:
  postgresdata:
```

**Comment by unixfox** [comment]
Please post the invidious version (bottom of the page) and check if you are up to date.

**Comment by TheEdgeOfRage** [comment]
I'm running the latest image, from 3 days ago

**Comment by unixfox** [comment]
Please post the version anyway thank you.

**Comment by swolford** [comment]
I'm chiming in with +1 on this issue - 403 on `videoplayback?` route. I'm having the same, except not as constantly.

Container logs say the same:
```
2023-08-11 06:44:55 UTC [info] 403 GET /videoplayback?expire=1691757657&ei=-dfVZM_TAvC9x_APvce1wA4&ip=REDACTED&id=o-AIiUCaC5r4PPv9NR8x2HCsd90e5z0F7lKnoyglgAp4Jx&itag=398&source=youtube&requiressl=yes&mh=jt&mm=31%2C29&mn=sn-32o-1the%2Csn-5hnekn7d&ms=au%2Crdu&mv=m&mvi=2&pl=26&initcwndbps=3022500&spc=UWF9f8N6p4VTO2oChYGdB3gnM1LBkKYXrWmuZCCZBw&vprv=1&svpuc=1&mime=video%2Fmp4&gir=yes&clen=46003166&dur=857.880&lmt=1691441992636539&mt=1691735604&fvip=5&keepalive=yes&fexp=24007246&c=ANDROID&txp=5537434&sparams=expire%2Cei%2Cip%2Cid%2Citag%2Csource%2Crequiressl%2Cspc%2Cvprv%2Csvpuc%2Cmime%2Cgir%2Cclen%2Cdur%2Clmt&sig=AOq0QJ8wRQIgKG7B8BX1QxXs7KG6F3AP5EBDGY9ZLpJYLMc4EWBsi5cCIQC_FtCjYUiBmORbDPliDnDvBgX8Ydim7OPtZj2Tjq2Vbg%3D%3D&lsparams=mh%2Cmm%2Cmn%2Cms%2Cmv%2Cmvi%2Cpl%2Cinitcwndbps&lsig=AG3C_xAwRQIgekKYcP9okaaUSpSym8cuxU-QEXgRj8hdYhUZEeTOGi4CIQCz6Yk5ChMi5Nke81ZAv6TyOhdakvNoMf71sRSVAAUksw%3D%3D&host=rr2---sn-32o-1the.googlevideo.com 30.69ms
```

Version: 
2023.08.07-3450896 @ master

Affected IDs:
umpuS0jJs-s
nQcPuFmA7XU
ZPSqhb4KKNc
Yi0WRF5WPFw

**Comment by TheEdgeOfRage** [comment]
@swolford funnily enough, all of those videos play on my instance. And whenever I encounter the 403 error, I just hit the "switch invidious instance" and they pretty much always work on another one.

**Comment by syeopite** [comment]
I'm guessing it might be a cache issue.

YouTube recently implemented a change that broke playback as various streaming urls returned will just yield a 403 instead of actual video content. See #4027 

A workaround was added in #4037 which was merged on August 7th. 

However, due to the video cache, streaming urls from before the fix are still being used. The aforementioned video working now is probably because the cache got cleared.

**Comment by TheEdgeOfRage** [comment]
I guess I'll leave this open for a couple days and see if the error resurfaces. If not, I'll close this issue

**Comment by blackandcold** [comment]
Rebuilding the Dockerimage and running docker-compose again made it work for now.
The patch was integrated 3 days ago.

**Comment by unixfox** [comment]
> I'm guessing it might be a cache issue.

I don't think so the cache is only kept for 6 hours max and refresh by invidious after that.

**Comment by foodornt** [comment]
I can affirm that this issue occurred, rebuilding (through `docker-compose build`) and restarting brought it back to normal

**Comment by TheEdgeOfRage** [comment]
I haven't encountered the issue at all in the past few days, so I'll close this as resolved.

**Comment by mortezadadgar** [comment]
I'm getting this error on https://iv.nboeck.de, https://vid.puffyan.us, https://invidious.io.lol, https://invidious.no-logs.com instances
youtube URL: https://www.youtube.com/watch?v=KtQZRAkgLqo
![Screenshot-230815-1416-55](https://github.com/iv-org/invidious/assets/27911493/6105d112-cdf0-416b-aaac-89903f51984d)

**Comment by davegallant** [comment]
I'm also hitting this more often than not.

I pulled the latest image today and restarted, but unfortunately it hasn't helped.

**Comment by unixfox** [comment]
Reopening as youtube appears to straight block the IP addresses of the popular public instances.

**Comment by blackandcold** [comment]
Worked two days somewhat and now no video can be played at all anymore.
Can I help with debugging somehow?

**Comment by unixfox** [comment]
The issue https://github.com/iv-org/invidious/issues/3915 still applies for discussing about how to circumvent youtube blockage.

This comment still apply for solutions on how to circumvent the youtube blockage: https://github.com/iv-org/invidious/issues/3822#issuecomment-1565401621

**Comment by gi-yt** [comment]
> The issue #3915 still applies for discussing about how to circumvent youtube blockage.
> 
> This comment still apply for solutions on how to circumvent the youtube blockage: [#3822 (comment)](https://github.com/iv-org/invidious/issues/3822#issuecomment-1565401621)

Youtube might be blocking subnets, I get a 403 on IPv6es in both the 16 IPv6 and routed subnets of BuyVM Luxembourg.
I also have the same issue on Racknerd Los Angeles,  but I think that is because of the v6 being in the AAAA record, I'll update here after getting a secondary v6 from them.

I also will try (re)setting up the ipv6 thing on my home network and trying with that, to see if that will help.

The interesting part in all this is that Piped continues to work, with all the IPv6es I threw at it after this wave of blockages..

**Comment by kuba-orlik** [comment]
Disabling IPV6 didn't help in my case

**Comment by Tha14** [comment]
I'd just like to add that I had the exact same problem with my instance and it went away after deleting the old docker image and building a new one. I guess that there was some breaking change on youtube's side and that invidious required a complete rebuild.

**Comment by unixfox** [comment]
> I'd just like to add that I had the exact same problem with my instance and it went away after deleting the old docker image and building a new one. I guess that there was some breaking change on youtube's side and that invidious required a complete rebuild.

It's very much possible that you were running an old invidious version that was blocked by youtube. We had to do some changes to get past this blockage.

It's unrelated to the current issue public instances are facing.

**Comment by ba06** [comment]
Something I've noticed is that certain livestreams (not live, but completed) seem to be failing. Every Invidious instance I can think of fails to load watch?v=lrqNLU-js-k which was streamed, as of writing, 11 hours ago. On instances where download is enabled, the only files that show up are either video only or audio only (and didn't download, either), so I'm guessing it's just Youtube having not finished processing the video.

I've noticed it happening a lot within the past 48-ish hours with completed livestreams, which is why I'm mentioning it here. Also live livestreams seem to have been broken for weeks if not months (JS on or off), but that could be Youtube just blocking the instance (I mainly use yewtu.be or vid.puffyan.us).

**Comment by absidue** [comment]
@ba06 Yes you are correct those completed live streams haven't been reprocessed to normal videos yet, how YouTube prioritises which ones get processed first, isn't really known but more popular ones tend to get processed sooner.

The player that Invidious (and FreeTube) use, video.js, can't handle YouTube's live stream DASH (technical jargon: self-initialising segments) but YouTube also provides HLS manifests for those post-live-DVR videos, it only has the last 2 hours of the live stream, but it's better than nothing.

TL;DR If Invidious treats those post live DVR videos as live streams, playback should work (check if `videoDetails.isPostLiveDvr` is `true` in the Innertube `/player` response).

**Comment by dr460nf1r3** [comment]
I suspect there must be a different issue but a particular IP is being blocked. For invidious.garudalinux.org as well as piped.garudalinux.org we are utilizing Squid to proxy outgoing requests via random IPs of our /64 subnet. Invidious is constantly giving 403 while Piped works just fine. Recently had to rotate addresses to keep it working. Invidious however still does not work. 

Restarting the instance is also not helping in this particular case. Is it possible that Invidious does not respect `HTTP(S)_PROXY` environment variables?

**Comment by gi-yt** [comment]
> I suspect there must be a different issue but a particular IP is being blocked. For invidious.garudalinux.org as well as piped.garudalinux.org we are utilizing Squid to proxy outgoing requests via random IPs of our /64 subnet. Invidious is constantly giving 403 while Piped works just fine. Recently had to rotate addresses to keep it working. Invidious however still does not work.
> 
> Restarting the instance is also not helping in this particular case. Is it possible that Invidious does not respect `HTTP(S)_PROXY` environment variables?

invidious does not support http proxies yet, there is WIP for it though: https://github.com/syeopite/invidious/tree/http-proxy

**Comment by dr460nf1r3** [comment]
Now that explains quite a lot. Thanks for the information!

**Comment by joaquinito2070** [comment]
I cannot play Disney Channel videos.

**Comment by unixfox** [comment]
Hello to all public instances owners or people that are running a private Invidious instance that is blocked!

We have officially released a tool that allow to automatically rotate the IPv6 address of your Invidious server: https://github.com/iv-org/smart-ipv6-rotator and thus escape from the YouTube block!

You can follow the detailed tutorial here: https://docs.invidious.io/ipv6-rotator/

Please give your feedback, it is very important to us!

**Comment by unixfox** [comment]
@gptlang someone on matrix said that they managed to force invidious to use ipv6 by adding this line in the config.yml:
```
host_binding: ::0
```

could you try?

**Comment by aram535** [comment]
I just installed invidious, via docker and none of the videos are working and giving this error.  Search works fine, I do not have proxy enabled.  I currently have a premium youtube subscription as well 

{
  "name": "invidious",
  "version": "2023.10.26-8087e64d",
  "branch": "master"
}

![image](https://github.com/iv-org/invidious/assets/3215912/38f1ab4a-cb1e-4c2d-9fd8-0d4879ed6b0b)

**Comment by unixfox** [comment]
> I just installed invidious, via docker and none of the videos are working and giving this error. Search works fine, I do not have proxy enabled. I currently have a premium youtube subscription as well
> 
> { "name": "invidious", "version": "2023.10.26-8087e64d", "branch": "master" }
> 
> ![image](https://user-images.githubusercontent.com/3215912/279311634-38f1ab4a-cb1e-4c2d-9fd8-0d4879ed6b0b.png)

Completely unrelated to this github issue. I can see that you have not correctly configured invidious all the way, please check this documentation: https://docs.invidious.io/installation/#post-install-configuration

**Comment by unixfox** [comment]
We have fixed an issue where invidious would not send all the traffic in ipv6: https://github.com/iv-org/invidious/pull/4225
Please add this line in your config.yml to be sure that invidious send the traffic in ipv6:
```yaml
force_resolve: ipv6
```

If you are using the rotator (https://docs.invidious.io/ipv6-rotator/), please make sure to have the line above and have updated Invidious!

**Comment by joaquinito2070** [comment]
403 in https://vid.puffyan.us/

**Comment by vk4s** [comment]
If you are self-hosting with docker and the videos stopped playing, updating the invidious docker image to `2024.04.26-eda7444-arm64` will fix that.

**Comment by Anjolao** [comment]
Hello.  My name is Anjola Adebayo.
The invidious is not working by saying, "This helps protect our community."
